sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2ScanRelationPushDown.scala

```diff
@@ -17,61 +17,193 @@
+                    case other => other.asInstanceOf[AttributeReference]
+                  }
+                  val output = groupAttrs ++ newOutput.drop(groupAttrs.length)
+
+                  logInfo(
+                    s"""
+                       |Pushing operators to ${sHolder.relation.name}
+                       |Pushed Aggregate Functions:
+                       | ${pushedAggregates.get.aggregateExpressions.mkString(", ")}
+                       |Pushed Group by:
+                       | ${pushedAggregates.get.groupByColumns.mkString(", ")}
+                       |Output: ${output.mkString(", ")}
+                      """.stripMargin)
+
+                  val scanRelation = DataSourceV2ScanRelation(sHolder.relation, scan, output)
+                  assert(scanRelation.output.length ==
+                    groupingExpressions.length + aggregates.length)
+
+                  val plan = Aggregate(
+                    output.take(groupingExpressions.length), resultExpressions, scanRelation)
+
+                  // scalastyle:off
+                  // Change the optimized logical plan to reflect the pushed down aggregate
+                  // e.g. TABLE t (c1 INT, c2 INT, c3 INT)
+                  // SELECT min(c1), max(c1) FROM t GROUP BY c2;
+                  // The original logical plan is
+                  // Aggregate [c2#10],[min(c1#9) AS min(c1)#17, max(c1#9) AS max(c1)#18]
+                  // +- RelationV2[c1#9, c2#10] ...
+                  //
+                  // After change the V2ScanRelation output to [c2#10, min(c1)#21, max(c1)#22]
+                  // we have the following
+                  // !Aggregate [c2#10], [min(c1#9) AS min(c1)#17, max(c1#9) AS max(c1)#18]
+                  // +- RelationV2[c2#10, min(c1)#21, max(c1)#22] ...
+                  //
+                  // We want to change it to
+                  // == Optimized Logical Plan ==
+                  // Aggregate [c2#10], [min(min(c1)#21) AS min(c1)#17, max(max(c1)#22) AS max(c1)#18]
+                  // +- RelationV2[c2#10, min(c1)#21, max(c1)#22] ...
+                  // scalastyle:on
+                  var i = 0
+                  val aggOutput = output.drop(groupAttrs.length)
+                  plan.transformExpressions {
+                    case agg: AggregateExpression =>
+                      i += 1
+                      val aggFunction: aggregate.AggregateFunction =
+                        agg.aggregateFunction match {
+                          case _: aggregate.Max => aggregate.Max(aggOutput(i - 1))
+                          case _: aggregate.Min => aggregate.Min(aggOutput(i - 1))
+                          case _: aggregate.Sum => aggregate.Sum(aggOutput(i - 1))
+                          case _: aggregate.Count => aggregate.Sum(aggOutput(i - 1))
+                          case _ => agg.aggregateFunction
+                        }
+                      agg.copy(aggregateFunction = aggFunction, filter = None)
```

Comment:
why do we need `filter = None`?

Reply:
When I initially prototyped with JDBC, I planned to handle the filter inside `AggregateExpression` by combining this filter with the regular filters, so I set it to `None` here. I will remove this `filter = None` since now we decide not to push down if the `AggregateExpression` has a filter.